# Module 2 - Data Inspection

## 1. Number of Rows

The number of rows represents the total number of records or observations present in a dataset.

In the Hotel Booking Demand dataset, each row represents one hotel booking record.

### Example

If a dataset contains 10,000 rows, it means there are 10,000 booking records available for analysis.

### Real-World Example

A hotel can use the number of records to understand how much booking history is available for analysis.

### Implementation

In [4]:
import pandas as pd

df = pd.read_csv("data/hotel_bookings.csv")

df.head()

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


In [26]:
print("Number of Rows:", df.shape[0])

Number of Rows: 119390


### Output

The output shows the total number of hotel booking records in the dataset.

### Interpretation

The output represents the total number of booking records available for analysis.

### Insight

Knowing the number of rows helps us understand the size of the dataset before performing further analysis.

### AI/ML Relevance

The number of observations is important in Machine Learning because it indicates how much data is available for learning patterns and building reliable models.

## 2. Number of Columns

The number of columns represents the total number of features or variables present in a dataset.

In the Hotel Booking Demand dataset, each column provides specific information about a hotel booking.

### Example

Columns can contain information such as hotel type, arrival date, number of guests, market segment, deposit type, and booking status.

### Real-World Example

A hotel can use the number of columns to understand how many different attributes are available for analyzing its bookings.

### Implementation

In [6]:
print("Number of Columns:", df.shape[1])

Number of Columns: 32


### Output

The output shows the total number of columns present in the dataset.

### Interpretation

Each column represents a variable or feature that describes some aspect of a hotel booking.

### Insight

The number of columns gives an initial idea of how much information is available for analysis.

### AI/ML Relevance

Understanding the number of features helps determine the input information available for Machine Learning and guides later feature selection and preprocessing.

## 3. Data Types

Data types describe the kind of values stored in each column.

Common data types include integers, floating-point numbers, strings, and Boolean values.

### Example

In the Hotel Booking Demand dataset, numerical columns such as adults and adr contain numeric values, while columns such as hotel and meal contain categorical or text values.

### Real-World Example

Knowing the data type helps a hotel analytics team understand how each column should be analyzed.

### Implementation

In [7]:
print(df.dtypes)

hotel                                 str
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                    str
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                                  str
country                               str
market_segment                        str
distribution_channel                  str
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                    str
assigned_room_type                    str
booking_changes                     int64
deposit_type                          str
agent                             

### Output

The output shows the data type of each column.

### Interpretation

Numerical columns can be used for calculations, while text-based columns generally represent categories or labels.

### Insight

Data types help us identify which columns may require further transformation during analysis or preprocessing.

### AI/ML Relevance

Machine Learning models require data to be represented in suitable formats, so understanding data types is an important preprocessing step.

## 4. Column Names

Column names identify the variables or features present in a dataset.

They help us understand what information each column represents.

### Example

The dataset contains columns related to hotel type, arrival information, customer details, booking information, and cancellation status.

### Real-World Example

A hotel analyst can use column names to quickly understand which information is available for each booking.

### Implementation

In [8]:
print(df.columns.tolist())

['hotel', 'is_canceled', 'lead_time', 'arrival_date_year', 'arrival_date_month', 'arrival_date_week_number', 'arrival_date_day_of_month', 'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children', 'babies', 'meal', 'country', 'market_segment', 'distribution_channel', 'is_repeated_guest', 'previous_cancellations', 'previous_bookings_not_canceled', 'reserved_room_type', 'assigned_room_type', 'booking_changes', 'deposit_type', 'agent', 'company', 'days_in_waiting_list', 'customer_type', 'adr', 'required_car_parking_spaces', 'total_of_special_requests', 'reservation_status', 'reservation_status_date']


### Output

The output displays all column names in the dataset.

### Interpretation

Each column name represents a specific variable related to hotel bookings.

### Insight

Reviewing column names helps us understand the available information before detailed analysis.

### AI/ML Relevance

Column names help identify potential features, target variables, and fields that may require further inspection.

## 5. Index

The index identifies the position of each row in a Pandas DataFrame.

By default, Pandas creates a numeric index starting from 0 when a custom index is not provided.

### Example

The first booking record has index 0, the second record has index 1, and so on.

### Real-World Example

An index allows analysts to locate and reference individual booking records within a DataFrame.

### Implementation

In [9]:
print(df.index)

RangeIndex(start=0, stop=119390, step=1)


### Output

The output shows the index information of the DataFrame.

### Interpretation

The index represents the row positions of the dataset.

### Insight

Understanding the index is useful when selecting, filtering, and referencing specific records.

### AI/ML Relevance

A correctly managed index helps maintain consistent row references during data preparation and analysis.

## 6. head()

The head() method displays the first few rows of a DataFrame.

By default, it displays the first five rows.

### Example

It can be used to quickly check whether the dataset was loaded correctly.

### Real-World Example

A hotel analyst can inspect the first few booking records before beginning detailed analysis.

### Implementation

In [12]:
print(df.head())

          hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1     

### Output

The first five rows of the dataset are displayed.

### Interpretation

The output gives an initial view of the records and their corresponding columns.

### Insight

 head() provides a quick way to understand the initial structure and content of a dataset.

### AI/ML Relevance

It is useful for verifying that the dataset has loaded correctly before starting preprocessing or model preparation.

## 7. tail()

The  tail()  method displays the last few rows of a DataFrame.

By default, it displays the last five rows.

### Example

It can be used to inspect the end of a dataset and check whether the records look consistent.

### Real-World Example

A hotel analyst can use tail() to review the latest or final booking records in the dataset.

### Implementation

In [18]:
print(df.tail())

             hotel  is_canceled  lead_time  arrival_date_year  \
119385  City Hotel            0         23               2017   
119386  City Hotel            0        102               2017   
119387  City Hotel            0         34               2017   
119388  City Hotel            0        109               2017   
119389  City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                        

### Output

The last five rows of the dataset are displayed.

### Interpretation

The output shows the records located at the end of the dataset.

### Insight

Checking the end of the dataset can help identify unexpected values or structural issues.

### AI/ML Relevance

Inspecting both the beginning and end of the dataset helps confirm that the data was loaded and stored correctly.

## 8. sample()

The sample()  method returns randomly selected rows from a DataFrame.

It is useful for getting a quick view of records from different parts of the dataset.

### Example

Instead of always viewing the first or last rows, sample() can show random booking records.

### Real-World Example

A hotel analyst can randomly inspect bookings to check whether different types of records are present throughout the dataset.

### Implementation

In [15]:
print(df.sample(5, random_state=42))

               hotel  is_canceled  lead_time  arrival_date_year  \
30946   Resort Hotel            0        203               2016   
40207     City Hotel            1         82               2015   
103708    City Hotel            0         25               2016   
85144     City Hotel            0          1               2016   
109991    City Hotel            0         70               2017   

       arrival_date_month  arrival_date_week_number  \
30946            December                        49   
40207                July                        29   
103708           December                        53   
85144               March                        11   
109991              April                        16   

        arrival_date_day_of_month  stays_in_weekend_nights  \
30946                           2                        2   
40207                          16                        0   
103708                         27                        0   
85144             

### Output

Five randomly selected booking records are displayed.

### Interpretation

The selected rows provide a quick look at records from different positions in the dataset.

### Insight

Random sampling can provide a broader view of the dataset than only examining the first or last rows.

### AI/ML Relevance

Sampling is useful for quickly checking large datasets and identifying potential data-quality issues before deeper analysis.

## 9. shape

The `shape` attribute returns the dimensions of a DataFrame in the form of rows and columns.

### Example

A result such as `(119390, 32)` means the dataset contains 119,390 rows and 32 columns.

### Real-World Example

A hotel analytics team can use the shape of the dataset to understand the size of the available booking data.

### Implementation

In [16]:
print("Dataset Shape:", df.shape)

Dataset Shape: (119390, 32)


### Output

The output shows the number of rows and columns as a tuple.

### Interpretation

The first value represents rows and the second value represents columns.

### Insight

The shape provides a quick summary of the overall dataset size.

### AI/ML Relevance

Dataset dimensions help determine the amount of training information and the number of available features.

## 10. info()
The info() method provides a summary of the DataFrame, including column names, non-null values, data types, and memory usage.

### Example

It provides a compact overview of the dataset structure.

### Real-World Example

A hotel analyst can use  info() to identify columns with missing values and understand the types of data stored in the dataset.

### Implementation

In [17]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  str    
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  str    
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal                       

### Output

The output displays the number of entries, column names, non-null counts, data types, and memory usage.

### Interpretation

Columns with fewer non-null values than the total number of rows contain missing values.

### Insight

 info()  helps identify data-quality issues and understand the structure of the dataset before detailed analysis.

### AI/ML Relevance

The information provided by info() helps determine which columns may require cleaning, conversion, or missing-value treatment before Machine Learning.

## 11. describe()

The  describe()  method provides descriptive statistics for numerical columns.

It includes count, mean, standard deviation, minimum, quartiles, and maximum values.

### Example

It can summarize numerical booking variables such as lead time, adults, and average daily rate.

### Real-World Example

A hotel can use descriptive statistics to understand typical booking values and the spread of numerical features.

### Implementation

In [19]:
print(df.describe())

         is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000            

### Output

The output displays descriptive statistics for the numerical columns.

### Interpretation

The statistics show the center, spread, and range of numerical variables.

### Insight

Descriptive statistics can reveal unusual ranges, large variations, and potential outliers that may require further analysis.

### AI/ML Relevance

These statistics help understand numerical features before preprocessing and Machine Learning model development.

## 12. dtypes

The  dtypes attribute returns the data type of each column in the DataFrame.

It provides a quick way to review the types of values stored in the dataset.

### Example

Numerical columns may have integer or floating-point types, while text-based columns may use string-related types.

### Real-World Example

A hotel analyst can use dtypes  to verify whether numerical and categorical fields are stored in suitable formats.

### Implementation

In [21]:
print(df.dtypes)

hotel                                 str
is_canceled                         int64
lead_time                           int64
arrival_date_year                   int64
arrival_date_month                    str
arrival_date_week_number            int64
arrival_date_day_of_month           int64
stays_in_weekend_nights             int64
stays_in_week_nights                int64
adults                              int64
children                          float64
babies                              int64
meal                                  str
country                               str
market_segment                        str
distribution_channel                  str
is_repeated_guest                   int64
previous_cancellations              int64
previous_bookings_not_canceled      int64
reserved_room_type                    str
assigned_room_type                    str
booking_changes                     int64
deposit_type                          str
agent                             

### Output

The output displays the data type associated with each column.

### Interpretation

The result helps distinguish numerical, categorical, and other column types.

### Insight

Reviewing data types helps identify columns that may need conversion or special handling.

### AI/ML Relevance

Correct data types are necessary for effective data preparation and Machine Learning workflows.

## 13. nunique()

The nunique()  method returns the number of unique values in each column.

It helps us understand the variety of values present in the dataset.

### Example

A column such as hotel may have only a small number of unique values, while other columns may contain many different values.

### Real-World Example

A hotel can use unique-value counts to understand how many different booking categories, meal types, or market segments exist.

### Implementation

In [23]:
print(df.nunique())

hotel                                2
is_canceled                          2
lead_time                          479
arrival_date_year                    3
arrival_date_month                  12
arrival_date_week_number            53
arrival_date_day_of_month           31
stays_in_weekend_nights             17
stays_in_week_nights                35
adults                              14
children                             5
babies                               5
meal                                 5
country                            177
market_segment                       8
distribution_channel                 5
is_repeated_guest                    2
previous_cancellations              15
previous_bookings_not_canceled      73
reserved_room_type                  10
assigned_room_type                  12
booking_changes                     21
deposit_type                         3
agent                              333
company                            352
days_in_waiting_list     

### Output

The output shows the number of unique values in each column.

### Interpretation

Columns with very few unique values are often categorical, while columns with many unique values may require further inspection.

### Insight

Unique-value counts help identify categorical variables, high-cardinality columns, and possible identifier-like fields.

### AI/ML Relevance

Understanding cardinality is important when selecting and encoding categorical features for Machine Learning.

## 14. value_counts()

The  value_counts()  method counts how frequently each unique value occurs in a column.

It is especially useful for analyzing categorical variables.

### Example

The distribution of hotel types or cancellation status can be analyzed using  value_counts().

### Real-World Example

A hotel can count how many bookings belong to each hotel type or how many bookings were canceled.

### Implementation

In [24]:
print(df["hotel"].value_counts())
print()
print(df["is_canceled"].value_counts())

hotel
City Hotel      79330
Resort Hotel    40060
Name: count, dtype: int64

is_canceled
0    75166
1    44224
Name: count, dtype: int64


### Output

The output shows the number of records belonging to each category in the selected columns.

### Interpretation

The counts show how frequently each category appears in the dataset.

### Insight

Value counts can reveal dominant categories and help identify possible imbalance in important variables such as the target.

### AI/ML Relevance

Understanding category distribution is important for identifying class imbalance and selecting suitable preprocessing techniques for Machine Learning.

## 15. Identify Important Column Types

After inspecting the dataset, we should identify numerical, categorical, date/time, identifier, and potential target columns.

### Numerical Columns

Numerical columns contain values that can be used for mathematical and statistical analysis.

### Categorical Columns

Categorical columns contain labels or groups.

### Date/Time Columns

Date/time columns contain information related to dates or booking time periods.

### Identifier Columns

Identifier columns are used to identify records or related entities and should be evaluated carefully before using them as predictive features.

### Potential Target Column

The potential target column is the variable that represents the outcome we want to understand or predict.

### Implementation

In [25]:
numerical_columns = df.select_dtypes(include="number").columns.tolist()
categorical_columns = df.select_dtypes(include="str").columns.tolist()

date_columns = [
    col for col in df.columns
    if "date" in col.lower()
]

target_column = "is_canceled"

print("Numerical Columns:", numerical_columns)
print("Categorical Columns:", categorical_columns)
print("Date/Time Columns:", date_columns)
print("Potential Target:", target_column)

Numerical Columns: ['is_canceled', 'lead_time', 'arrival_date_year', 'arrival_date_week_number', 'arrival_date_day_of_month', 'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children', 'babies', 'is_repeated_guest', 'previous_cancellations', 'previous_bookings_not_canceled', 'booking_changes', 'agent', 'company', 'days_in_waiting_list', 'adr', 'required_car_parking_spaces', 'total_of_special_requests']
Categorical Columns: ['hotel', 'arrival_date_month', 'meal', 'country', 'market_segment', 'distribution_channel', 'reserved_room_type', 'assigned_room_type', 'deposit_type', 'customer_type', 'reservation_status', 'reservation_status_date']
Date/Time Columns: ['arrival_date_year', 'arrival_date_month', 'arrival_date_week_number', 'arrival_date_day_of_month', 'reservation_status_date']
Potential Target: is_canceled


### Interpretation

The identified columns provide an initial classification of the dataset structure.

Numerical columns can be analyzed statistically, categorical columns can be analyzed by category, date-related columns can be examined for time-based patterns, and the target column represents the main outcome for analysis.

### Insight

This inspection helps us understand the dataset before moving to data-quality assessment and Exploratory Data Analysis.

### AI/ML Relevance

Correctly identifying feature types and the target variable is essential for choosing appropriate preprocessing, analysis, visualization, and Machine Learning techniques.

## Data Inspection Summary

The dataset has now been inspected using its basic structure, dimensions, data types, column names, index, sample records, descriptive statistics, unique-value counts, and category frequencies.

The next step is to investigate data-quality issues such as missing values, duplicate records, invalid values, and unexpected data types.